feat(core): Coordinate manual workflow activation and deactivation in multi-main scenario

ivov opened the pull request:
Followup to #7566 | Story: https://linear.app/n8n/issue/PAY-926

### Manual workflow activation and deactivation

In a multi-main scenario, if the user manually activates or deactivates a workflow, the process (whether leader or follower) that handles the PATCH request and updates its internal state should send a message into the command channel, so that all other main processes update their internal state accordingly:

- Add to `ActiveWorkflows` if activating
- Remove from `ActiveWorkflows` if deactivating
- Remove and re-add to `ActiveWorkflows` if the update did not change activation status.

After updating their internal state, if activating or deactivating, the recipient main processes should push a message to all connected frontends so that these can update their stores and so reflect the value in the UI.

### Workflow activation errors

On failure to activate a workflow, the main instance should record the error in Redis - main instances should always pull activation errors from Redis in a multi-main scenario.

### Leadership change

On leadership change...

- The old leader should stop pruning and the new leader should start pruning.
- The old leader should remove trigger- and poller-based workflows and the new leader should add them.

ivov committed:
Handle workflow activation if leader

ivov committed:
Centralize into `MultiMainSetup`

ivov committed:
Broadcast workflow update event

ivov committed:
Fix license test

ivov committed:
Merge master

ivov committed:
Fix pruning test

ivov committed:
Add missing mocks

ivov committed:
Clean up tests

ivov committed:
Merge master

ivov committed:
Inform FE of activation and deactivation

ivov committed:
Fix orchestration test

krynble reviewed (commented):
Good job overall! Left a few comments. Looking good!

Still haven't checked the tests as you said some are broken.

ivov committed:
Merge master, fix conflicts

ivov committed:
Fix some tests

ivov commented on packages/cli/src/commands/start.ts:
```diff
@@ -233,32 +232,25 @@ export class Start extends BaseCommand {
-			!Container.get(License).isMultipleMainInstancesLicensed()
-		) {
+		if (multiMainSetup.isLeader && !Container.get(License).isMultipleMainInstancesLicensed()) {
 			throw new FeatureNotLicensedError(LICENSE_FEATURES.MULTIPLE_MAIN_INSTANCES);
```
This is correct, I'll keep it in mind when writing the docs.

ivov committed:
Simplify init in `BaseCommand`

ivov committed:
Remove leftover logs

ivov commented on packages/cli/src/services/orchestration/main/handleCommandMessageMain.ts:
```diff
@@ -60,10 +72,86 @@ export async function handleCommandMessageMain(messageString: string) {
+
+				const { workflowId, sessionId } = message.payload ?? {};
+
+				console.log('[activation] sessionId', sessionId);
```
> Maybe remove this console.log or replace it by a debug message? Same below.

Removed, were leftovers.

> If we decide to make this call "blocking"

Got it. Based on our discussion, we'll defer the decision of whether or not to block for the next PR.

ivov committed:
Remove unneeded handlers

ivov committed:
Fix webhooks tests

ivov committed:
Refactor pruning in multi-main scenario

flipswitchingmonkey commented on packages/cli/src/services/orchestration/main/MultiMainSetup.ee.ts:
```diff
@@ -89,4 +97,32 @@ export class MultiMainInstancePublisher extends SingleMainInstancePublisher {
+		});
+	}
+
+	async broadcastWorkflowWasActivated(workflowId: string, targets: string[], pushSessionId = '') {
+		if (!this.sanityCheck()) return;
+
+		await this.redisPublisher.publishToCommandChannel({
+			command: 'workflowWasActivated',
+			targets,
+			payload: { workflowId, pushSessionId },
+		});
+	}
+
+	async broadcastWorkflowWasDeactivated(workflowId: string, pushSessionId = '') {
+		if (!this.sanityCheck()) return;
+
+		await this.redisPublisher.publishToCommandChannel({
+			command: 'workflowWasDeactivated',
+			payload: { workflowId, pushSessionId },
+		});
+	}
```
We could extend the sanityCheck() with more checks, though I'm not sure what we could add. We are currently testing for queue mode and whether it is main, but only the Leader should be sending these out since otherwise we can start an endless loop when the receivers start reactivating theirs...?

ivov committed:
Fix `Start` command test

ivov commented on packages/cli/src/services/orchestration/main/MultiMainSetup.ee.ts:
```diff
@@ -1,19 +1,27 @@
-export class MultiMainInstancePublisher extends SingleMainInstancePublisher {
+export class MultiMainSetup extends SingleMainSetup {
+	isEnabled =
+		config.getEnv('executions.mode') === 'queue' && config.getEnv('leaderSelection.enabled');
```
I'll think about this, `License` already depends on `MultiMainSetup` so we'd create a cycle.

flipswitchingmonkey commented on packages/cli/src/ActiveWorkflowRunner.ts:
```diff
@@ -940,6 +943,17 @@ export class ActiveWorkflowRunner implements IWebhookManager {
+				'[Multi-main setup] Instance is follower, skipping removal of workflow from active workflows in memory and broadcasting "workflowWasDeactivated" into command channel...',
+			);
+
+			await this.multiMainSetup.broadcastWorkflowWasDeactivated(workflowId);
```
What should the reaction by the listening clients be when receiving this message? Iirc the idea was that when some client receives this, it also deactivates that workflow. But that means this message should only every been sent once, by the leader, not by followers, or am I mistaken here?

ivov replied:
Solved during pairing

flipswitchingmonkey reviewed (dismissed):
let's discuss the communication protocol and what should happen when messages are received (and who should send what when)

ivov committed:
Merge master

ivov committed:
Fix tests finally

ivov committed:
Add more missing mocks

ivov committed:
Break dep cycle

ivov committed:
Fix `webhooks.api.test.ts`

ivov committed:
Fix relabeling on destructuring

ivov committed:
Refactor to use `workflowActiveStateChanged`

ivov committed:
Forgotten removal

ivov committed:
Coordinate activation errors

ivov committed:
Fix tests

ivov committed:
Refactor to global config

ivov committed:
Remove unneeded leader check

This ensures that on license expiration, after 72 hours at most, no
further followers can be started.

ivov committed:
Merge master, fix conflicts

ivov committed:
Fix misresolved conflict

ivov committed:
Prefix context for pruning log messages

ivov committed:
Stop pruning if follower

ivov committed:
Implement activation errors in Redis

ivov committed:
Fix lint

ivov committed:
Fix import

ivov committed:
Remove TODOs

ivov committed:
Remove unused method and type

ivov committed:
Better naming

ivov committed:
Call `toCacheKey` only if needed

ivov committed:
Fix lint

ivov committed:
Add guard for init

ivov committed:
Better naming

ivov committed:
Rename service file

ivov committed:
Simplify error creation

ivov committed:
Add more guards

ivov committed:
Cleanup

ivov committed:
Tighten guard

ivov committed:
Add clarifying comment

ivov committed:
More cleanup

ivov committed:
Better naming

ivov committed:
Add defensive `init`

ivov committed:
Add guard to `WorkflowService`

ivov committed:
Cleanup

ivov commented:
@krynble @flipswitchingmonkey Submitting a bit faster than I'd like to as I need to go back to cloud support, but this PR should be 95% there and ready for review.

krynble commented on packages/cli/src/services/orchestration/main/MultiMainSetup.ee.ts:
```diff
@@ -69,24 +80,44 @@ export class MultiMainInstancePublisher extends SingleMainInstancePublisher {
+		if (!this.sanityCheck()) return;
+
+		await this.redisPublisher.publishToCommandChannel({
+			command: 'workflowActiveStateChanged',
```
```suggestion
			command: 'reloadWorkflowActiveState',
```

Based on the conversation yesterday, commands should indicate imperative actions.

ivov replied:
From what I recall, what we decided yesterday was that these were not commands but rather events, so it'd be the others (and methods, etc.) that need renaming. 

In any case, I tend to think we should avoid a single one being inconsistent with the others - ideally we could rename all of this implementation in a standalone PR.

krynble commented on packages/cli/src/commands/start.ts:
```diff
@@ -231,38 +228,32 @@ export class Start extends BaseCommand {
+		await multiMainSetup.init();
+
+		multiMainSetup.on('leadershipChange', async () => {
+			if (multiMainSetup.isLeader) {
```
Inside this `if` block, since this means that the current instance just became the leader, we should clear all previously existing `ActivationErrors` so that this instance would then report its own errors, if any happen, when initializing triggers wdyt?

ivov replied:
Addressed here: [7435dd1](https://github.com/n8n-io/n8n/pull/7643/commits/7435dd10135be87c898668cb1ead7f9633a5c7aa)

krynble commented on packages/editor-ui/src/Interface.ts:
```diff
@@ -419,7 +419,19 @@ export type IPushData =
 	| PushDataExecutionRecovered
-	| PushDataWorkerStatusMessage;
+	| PushDataWorkerStatusMessage
+	| PushDataActiveWorkflowAdded
+	| PushDataActiveWorkflowRemoved;
+
+type PushDataActiveWorkflowAdded = {
+	data: IActiveWorkflowAdded;
+	type: 'workflowActivated';
+};
+
+type PushDataActiveWorkflowRemoved = {
+	data: IActiveWorkflowRemoved;
+	type: 'workflowDeactivated';
+};
```
We might not need these anymore right? This was about "multiplayer real time update" - I guess we removed this, didn't we?

krynble commented on packages/editor-ui/src/mixins/pushConnection.ts:
```diff
@@ -291,6 +291,16 @@ export const pushConnection = defineComponent({
 				}
 			}
 
+			if (receivedData.type === 'workflowActivated') {
+				this.workflowsStore.setWorkflowActive(receivedData.data.workflowId);
+				return true;
+			}
+
+			if (receivedData.type === 'workflowDeactivated') {
+				this.workflowsStore.setWorkflowInactive(receivedData.data.workflowId);
+				return true;
+			}
+
```
Might also want to remove this.

ivov replied:
See above: https://github.com/n8n-io/n8n/pull/7643#discussion_r1393990223

krynble commented on packages/editor-ui/src/stores/workflows.store.ts:
```diff
@@ -551,6 +551,9 @@ export const useWorkflowsStore = defineStore(STORES.WORKFLOWS, {
 			if (this.workflowsById[workflowId]) {
 				this.workflowsById[workflowId].active = true;
 			}
+			if (workflowId === this.workflow.id) {
+				this.setActive(true);
+			}
```
And the changes in here perhaps?

ivov replied:
See above: https://github.com/n8n-io/n8n/pull/7643#discussion_r1393990223

ivov committed:
Replace `ActivationError` with string

ivov commented on packages/cli/src/ActivationErrors.service.ts:
```diff
@@ -0,0 +1,40 @@
+export class ActivationErrorsService {
+	constructor(private readonly cacheService: CacheService) {}
+
+	async set(workflowId: string, error: ActivationError) {
```
Thanks! Noticed the time was unused but thought of this as tech debt. Addressed.

ivov commented on packages/cli/src/ActiveWorkflowRunner.ts:
```diff
@@ -105,17 +92,13 @@ export class ActiveWorkflowRunner implements IWebhookManager {
-
-			await this.multiMainInstancePublisher.init();
+		if (config.getEnv('executions.mode') === 'queue' && config.getEnv('multiMainSetup.enabled')) {
+			await this.multiMainSetup.init();
```
Agree! This was more defensive than anything.

ivov committed:
Add missing check

ivov committed:
Fix lint

ivov committed:
Ensure also followers init license

ivov commented on packages/cli/src/commands/BaseCommand.ts:
```diff
@@ -243,19 +243,15 @@ export abstract class BaseCommand extends Command {
+			config.getEnv('multiMainSetup.instanceType') === 'follower'
+		) {
+			this.logger.debug(
+				'[Multi-main setup] Instance is follower, skipping license initialization...',
```
Addressed here: [7435dd1](https://github.com/n8n-io/n8n/pull/7643/commits/7435dd10135be87c898668cb1ead7f9633a5c7aa)

ivov committed:
Consolidate activation errors into single object

ivov committed:
Remove no longer relevant test

krynble commented on packages/cli/src/ActivationErrors.service.ts:
```diff
@@ -0,0 +1,52 @@
+	constructor(private readonly cacheService: CacheService) {}
+
+	async set(workflowId: string, errorMessage: string) {
+		const errors = await this.getAll();
+
+		errors[workflowId] = errorMessage;
+
+		await this.cacheService.set(this.cacheKey, JSON.stringify(errors));
```
Thanks for addressing this to use a single key - but I think we might have a bigger problem now.

See, on line 16, we perform `getAll`. When we talk about workflow activation, we nearly try to activate all workflows at the same time, right? So if 2 workflows fail, we could end up in a scenario like this:

- Error for workflow A performs `getAll` - it returns an empty list
- Error for workflow B also performs `getAll` - it also gets an empty list. The two calls are cocnurrent.
- Error from workflow A is set, like `{"workflowA": "Trigger failed"}`
- Error from workflow B is set, like `{"workflowB": "Trigger failed"}`
- Error from workflow A is saved; Redis contains `{"workflowA": "Trigger failed"}`
- Error from workflow B is then saved, overriding what A just did, and now Redis contains only `{"workflowB": "Trigger failed"}`

To address this I think it's best to improve the `cacheService` to allow for atomic hash operations using [Redis Hashes](https://redis.io/docs/data-types/hashes/). WDYT?

ivov replied:
Interesting!

> When we talk about workflow activation, we nearly try to activate all workflows at the same time, right?

So activations are sequential: in `ActiveWorkflowRunner.addActiveWorkflows()` we loop over them and on each iteration we call `ActiveWorkflowRunner.add()` which sets the error via `ActiveErrorsService.set()` which internally calls `ActiveErrorsService.getAll()`. If they are sequential, your scenario might not be an issue during activations?

> improve the cacheService to allow for atomic hash operations using [Redis Hashes](https://redis.io/docs/data-types/hashes/)

At first glance the caching packages we're using don't seem to support Redis hashes? I might be wrong - I'll take a look.

krynble replied:
You seem to be right, the lib we picked does not seem to support this. 

In terms of sequential activation, this seems right, so not as bad, but still possible. It'd be great if we could prevent this loophole but it doesn't seem possible with the current approach.

ivov replied:
Given this would require switching libs, we agreed this could be addressed in future but it should be fine for now.

krynble reviewed (commented):
One last comment from me.

ivov committed:
Add telemetry

ivov committed:
fix(core): Account for activation error caused by follower `main` instance (#7735)

This PR accounts for the edge case where the user on a FE connected to a
follower `main` instance leads to an activation failure. The follower
performs the DB update, skips activation in `ActiveWorkflows`, instructs
leader to perform activation in `ActiveWorkflows`, leader fails to
activate and reverts the DB update and communicates activation error to
follower, which broadcasts the event to all connected FEs, which update
their stores and display the activation error.

To reproduce, create a workflow with a trigger with an invalid
credential and attempt to activate on both leader and follower to
trigger the activation error - behavior should be identical on both.

krynble commented on packages/editor-ui/src/mixins/pushConnection.ts:
```diff
@@ -291,6 +291,30 @@ export const pushConnection = defineComponent({
+				this.showError(
+					new Error(receivedData.data.errorMessage),
+					this.$locale.baseText('workflowActivator.showError.title', {
+						interpolate: { newStateName: 'deactivated' },
```
```suggestion
						interpolate: { newStateName: 'activated' },
```

The modal title is currently shown as `Workflow could not be deactivated:` which seems a bit misleading.

ivov replied:
Please see [a3147a9](https://github.com/n8n-io/n8n/pull/7643/commits/a3147a926a9501bf373b0bf5a63717f17fca97e8)

krynble commented on packages/editor-ui/src/mixins/pushConnection.ts:
```diff
@@ -291,6 +291,30 @@ export const pushConnection = defineComponent({
+				this.workflowsStore.setWorkflowInactive(receivedData.data.workflowId);
+				this.workflowsStore.setActive(false);
+
+				this.showError(
```
We should add an if to check whether the currently visible workflow ID matches the one we just received. The way this is currently working, I tested with 3 main instances and if I click the toggle in one of the followers, the other displays the error even if another totally unrelated workflow is open.

ivov replied:
Please see [a3147a9](https://github.com/n8n-io/n8n/pull/7643/commits/a3147a926a9501bf373b0bf5a63717f17fca97e8)

krynble reviewed (commented):
Looking great, two very tiny changes

ivov committed:
Address feedback

krynble reviewed (approved):
Amazing work, loved it!

janober commented:
Got released with [`n8n@1.18.0`](https://github.com/n8n-io/n8n/releases/tag/n8n%401.18.0)
